libs/base/Control/Relation.idr

```diff
@@ -0,0 +1,79 @@
+||| Properties of binary relations
+module Control.Relation
+
+%default total
+
+public export
+interface Reflexive ty rel where
+  reflexive : {x : ty} -> rel x x
+
+public export
+interface Transitive ty rel where
+  transitive : {x, y, z : ty} -> rel x y -> rel y z -> rel x z
+
+public export
+interface Symmetric ty rel where
+  symmetric : {x, y : ty} -> rel x y -> rel y x
+
+public export
+interface Antisymmetric ty rel where
+  antisymmetric : {x, y : ty} -> rel x y -> rel y x -> x = y
+
+public export
+interface Dense ty rel where
+  dense : {x, y : ty} -> rel x y -> (z : ty ** (rel x z, rel z y))
+
+public export
+interface Serial ty rel where
+  serial : {x : ty} -> (y : ty ** rel x y)
+
+public export
+interface Euclidean ty rel where
+  euclidean : {x, y, z : ty} -> rel x y -> rel x z -> rel y z
+
+public export
+interface (Reflexive ty rel, Symmetric ty rel) => Tolerance ty rel where
+
+public export
+interface (Reflexive ty rel, Transitive ty rel, Symmetric ty rel) => Equivalence ty rel where
+
+----------------------------------------
+
+public export
+Reflexive ty rel => Dense ty rel where
+  dense {x} xy = (x ** (reflexive {x}, xy))
+
+public export
+Reflexive ty rel => Serial ty rel where
+  serial {x} = (x ** reflexive {x})
+
+public export
+(Transitive ty rel, Symmetric ty rel, Serial ty rel) => Reflexive ty rel where
+  reflexive {x} =
+    let (y ** xy) = serial {x} in
+      transitive {x} xy $ symmetric {x} xy
+
+public export
+(Reflexive ty rel, Euclidean ty rel) => Symmetric ty rel where
+  symmetric {x} xy = euclidean {x} xy $ reflexive {x}
+
+public export
+(Reflexive ty rel, Euclidean ty rel) => Transitive ty rel where
+  transitive {rel} xy yz = symmetric {rel} $ euclidean {rel} yz $ symmetric {rel} xy
+
+public export
+(Transitive ty rel, Symmetric ty rel) => Euclidean ty rel where
+  euclidean {rel} xy xz = transitive {rel} (symmetric {rel} xy) xz
```

Comment:
Should these be named? It feels like they'd get in the way during instance resolution.

Reply:
That sounds like a good idea, but I don't know how to do it. There were also some other relation laws that I wanted to added, but I couldn't get the compiler to accept them all. I found the compiler to be very finicky when it came to this part.

Reply:
You can write `[TSE] (Transitive ty rel, Symmetric ty rel) => Euclidean ty rel where (...)`
to name the instance `[TSE]`. You should then be able to use it in specific instance declarations
by writing `using TSE`, cf. the existing work on additive vs. multiplicative semigroups:

https://github.com/idris-lang/Idris2/blob/9c63e90fd215526c2e7a8a83cbb6ad6aa2e3f62c/libs/contrib/Data/Nat/Algebra.idr#L10-L12

Reply:
Okay, I got them named. One problem left, which is that I don't know how to use these proofs. For example, it's easy to prove that `Equal` is transitive and symmetric. It should be possible to use those facts along with `TSE` to prove it's also euclidean, but I don't know how to do that.

Reply:
Hmmm. I guess there isn't good syntax for it. We may want to open an issue for this.
Here are two possible solutions, each with their inconvenients:

```idris
%hint EqualEuclidean : Euclidean ty Equal
EqualEuclidean = TSE

Euclidean ty Equal where
  euclidean = euclidean {rel = Equal} @{TSE}
```

Reply:
That did it! The hint was not required.

Reply:
The `%hint` is so that it actually behaves like an implementation and is used by auto search!